frame/collective/src/lib.rs

```diff
@@ -168,12 +175,129 @@ decl_error! {
 		AlreadyInitialized,
 		/// The close call is made too early, before the end of the voting.
 		TooEarly,
+		/// There can only be a maximum of `MaxMembers`.
+		TooManyMembers,
+		/// There can only be a maximum of `MaxProposals` active proposals.
+		TooManyProposals,
+		/// The given weight bound for the proposal was too low.
+		WrongProposalBound,
+	}
+}
+
+/// Functions for calcuating the weight of dispatchables.
+mod weight_for {
+	use frame_support::weights::{RuntimeDbWeight, Weight};
+
+	/// Calculate the weight for `set_members`.
+	pub(crate) fn set_members(
+		db: RuntimeDbWeight,
+		old_count: impl Into<Weight>,
+		new_count: impl Into<Weight>,
+		proposals: impl Into<Weight> + Copy,
+	) -> Weight {
+		db.reads_writes(1, 1) // mutate `Members`
+			.saturating_add(db.writes(1)) // set `Prime`
+			.saturating_add(db.reads(1)) // read `Proposals`
+			.saturating_add(db.reads_writes(proposals.into(), proposals.into())) // update votes (`Voting`)
+			.saturating_add(old_count.into().saturating_mul(21_000_000)) // M
+			.saturating_add(new_count.into().saturating_mul(110_000)) // N
+			.saturating_add(proposals.into().saturating_mul(27_000_000)) // P
+	}
+
+	// The bytes are responsible for an insignificant amount of weight.
+	// Length of proposals is limited by length fees.
+	const MAX_ASSUMED_PROPOSAL_BYTES: Weight = 1_024;
```

Comment:
I do not believe this is a reasonable assumption.

I am looking at the code, weight introduced by the byte length seems entirely caused by hashing the proposal.

And that hash only exists for depositing an event. I would ask if we can just remove all that logic. Why do we need the proposal hash for a member executed function?

Reply:
If this is not the case, we should find someone who can help pass the proposal length higher up the stack where it is cheap to do so.

Reply:
I imagine that it's a sanity check? You could just give out the id/index, but the hash is to make sure that you have the right proposal?

Reply:
One solution can be to ask the encoded_proposal in parameter and decode it inside the extrinsic. (this is more efficient as we no longer encode it again in order to hash it)